middleware.js

```diff
@@ -0,0 +1,211 @@
+                    return new Promise(function (resolve, reject) {
+                        var middlewarePromise = _self._getMiddlewarePromise(currentMiddleware, params, value);
+                        return middlewarePromise.then(
+                            function onResolved() {
+                                // resolve the promise in order to continue with main flow
+                                resolve.apply(this, arguments);
+                            },
+                            function onRejected(error) {
+                                if (typeof currentMiddleware.errorHandler === 'function') {
+                                    try {
+                                        /** if there is an errorHandler, call the resolve with returned value of the errorHandler
+                                         in order to continue with the main flow */
+                                        resolve(params, currentMiddleware.errorHandler(error));
+                                    }
+                                    catch (e) {
+                                        /** if there is an error in the middleware errorHandler, ejects the promises and stop the execution  */
+                                        reject(e);
+                                    }
+                                } else {
+                                    /** if there is not an errorHandler, rejects the promises and stop the execution **/
+                                    reject(error);
+                                }
+                            }
+                        );
+                    });
+                });
+            }, Promise.resolve());
+		},
+
+
+		/**
+		 * Works as a wrapperExecutes, executes the middleware.run function and always returns a promise even if the
+		 * run method is promise or a regular function
+		 * @param {Object} middleware - middleware to run
+		 * @param {Object?} params - run method params
+		 * @param value - previous promise returned value
+         * @returns {Promise}
+         * @private
+         */
+		_getMiddlewarePromise : function (middleware,params ,value) {
+			var middlewarePromise;
+			try{
+				middlewarePromise = middleware.run(params, value);
+				if (middlewarePromise && typeof middlewarePromise.then === 'function'){
+					return middlewarePromise;
+				}else{
+					if (middlewarePromise && !(middlewarePromise instanceof Error)){
+						return Promise.resolve(middlewarePromise);
+					}else
+					{
+						return Promise.reject(middlewarePromise);
+					}
+				}
+			}catch(error){
+				return Promise.reject(error);
+			}
+		}
+	});
+
+
+
+	Middleware.classMembers({
+		/**
+		 * @constant {Object} PHASES - enum to define all the phases
+		 */
+		PHASES : {
+			ROUTE : 'route'
+		},
+
+        /**
+         * @constant {Object} SUBPHASES - enum to define all the sub phases
+         */
+		SUBPHASES : {
+			BEFORE : 'before',
+			DURING : 'during',
```

Comment:
shouldn't we have here the constant `DEFAULT_SUBPHASE` to ensure they're the same?